session_resync.go

```diff
@@ -0,0 +1,79 @@
+package remotedialer
+
+import (
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"time"
+)
+
+var errCloseResync = errors.New("resync from client")
+
+// encodeConnectionIDs serializes a slice of connection IDs
+func encodeConnectionIDs(ids []int64) []byte {
+	payload := make([]byte, 0, 8*len(ids))
+	for _, id := range ids {
+		payload = binary.LittleEndian.AppendUint64(payload, uint64(id))
+	}
+	return payload
+}
+
+// decodeConnectionIDs deserializes a slice of connection IDs
+func decodeConnectionIDs(payload []byte) ([]int64, error) {
+	if len(payload)%8 != 0 {
+		return nil, fmt.Errorf("incorrect data format")
+	}
+	result := make([]int64, 0, len(payload)/8)
+	for x := 0; x < len(payload); x += 8 {
+		id := binary.LittleEndian.Uint64(payload[x : x+8])
+		result = append(result, int64(id))
+	}
+	return result, nil
+}
+
+func newResyncMessage(connectionIDs []int64) *message {
+	return &message{
+		id:          nextid(),
+		messageType: Resync,
+		bytes:       encodeConnectionIDs(connectionIDs),
+	}
+}
+
+// sendResync sends a binary message of type Resync, whose payload is a list of the active connection IDs for this session
+func (s *Session) sendResync() error {
+	_, err := s.writeMessage(time.Now().Add(ResyncTimeout), newResyncMessage(s.activeConnectionIDs()))
+	return err
+}
+
+// lockedResyncConnections closes any session connection that is not present in the IDs received from the client
+func (s *Session) lockedResyncConnections(payload []byte) error {
+	clientIDs, err := decodeConnectionIDs(payload)
+	if err != nil {
+		return fmt.Errorf("decoding resync connections payload: %w", err)
+	}
+	for _, id := range s.activeConnectionIDs() {
+		if i := sliceIndex(clientIDs, id); i >= 0 {
+			// Connection is still active, do nothing
+			clientIDs = append(clientIDs[:i], clientIDs[i+1:]...) // IDs are unique, skip in next iteration
```

Comment:
Given we have two ordered lists (`activeConnectionIDs()` and `clientIDs`), how about using the classic merge-sort-like algorithm?

You keep two indices over the two lists, both starting at 0. Then you bump one or the other index depending on which list has the element with the greatest value. At each step you collect in a result list elements that are in the first but not in the second list.

Reply:
Done at e7efd38, thanks!